### PR TITLE
Support for regression in YOLOv8 framework

### PR DESCRIPTION
**Pull request by nklimov23**

### CODE REVIEW

File: ultralytics/cfg/__init__.py [hunk 1]
```diff
@@ -31,13 +31,14 @@
...
 MODES = "train", "val", "predict", "export", "track", "benchmark"
-TASKS = "detect", "segment", "classify", "pose", "obb"
+TASKS = "detect", "segment", "classify", "pose", "obb", "regress"
 TASK2DATA = {
     "detect": "coco8.yaml",
     "segment": "coco8-seg.yaml",
     "classify": "imagenet10",
     "pose": "coco8-pose.yaml",
     "obb": "dota8.yaml",
+    "regress": "imdb-age.yaml"
```
Comment (khatami-mehrdad) [inline ultralytics/cfg/__init__.py]: These datasets are used for verification.
Do you have a small version of "imdb-age.yaml" to replace this?

File: ultralytics/data/augment.py [hunk 2]
```diff
@@ -1244,6 +1244,9 @@ def __call__(self, im):
             (torch.Tensor): The transformed image as a PyTorch tensor in float32 or float16, normalized to [0, 1].
         """
         im = np.ascontiguousarray(im.transpose((2, 0, 1))[::-1])  # HWC to CHW -> BGR to RGB -> contiguous
+        #im = im.astype(np.float32())
```
Comment (khatami-mehrdad) [inline ultralytics/data/augment.py]: Any reason for these comments.
can you remove these commented lines?

File: ultralytics/engine/exporter.py [hunk 3]
```diff
@@ -460,8 +464,16 @@ def transform_fn(data_item):
 
             # Generate calibration data for integer quantization
             LOGGER.info(f"{prefix} collecting INT8 calibration images from 'data={self.args.data}'")
-            data = check_det_dataset(self.args.data)
-            dataset = YOLODataset(data["val"], data=data, imgsz=self.imgsz[0], augment=False)
+            if self.model.task in ("classify", "regress"):
```
Comment (khatami-mehrdad) [inline ultralytics/engine/exporter.py]: can you write this in three ifs
if self.model.task == "classify":

elif self.model.task == "regress"

else
--

File: ultralytics/engine/exporter.py [hunk 4]
```diff
@@ -761,8 +774,16 @@ def export_saved_model(self, prefix=colorstr("TensorFlow SavedModel:")):
             if self.args.data:
                 # Generate calibration data for integer quantization
                 LOGGER.info(f"{prefix} collecting INT8 calibration images from 'data={self.args.data}'")
-                data = check_det_dataset(self.args.data)
-                dataset = YOLODataset(data["val"], data=data, imgsz=self.imgsz[0], augment=False)
+                if self.model.task in ("classify", "regress"):
```
Comment (khatami-mehrdad) [inline ultralytics/engine/exporter.py]: can you write this in three ifs
if self.model.task == "classify":

elif self.model.task == "regress"

else

File: ultralytics/models/yolo/classify/val.py [hunk 5]
```diff
@@ -29,6 +29,7 @@ class ClassificationValidator(BaseValidator):
     def __init__(self, dataloader=None, save_dir=None, pbar=None, args=None, _callbacks=None):
```
Comment (khatami-mehrdad) [inline ultralytics/models/yolo/classify/val.py]: The reason you changed classify is due to regress being based on classify?

File: ultralytics/models/yolo/pose/val.py [hunk 6]
```diff
@@ -64,39 +64,51 @@ def get_desc(self):
         )
 
     def postprocess(self, preds, img_shape):
-        """Apply non-maximum suppression and return detections with high confidence scores."""
         if self.separate_outputs:  # Quant friendly export with separated outputs
-            pred_order, nkpt = separate_outputs_decode(preds, self.args.task)
+            #mcv = float('-inf')
```
Comment (khatami-mehrdad) [inline ultralytics/models/yolo/pose/val.py]: ???? Why have you changed pose val

File: ultralytics/nn/autobackend.py [hunk 7]
```diff
@@ -334,6 +334,8 @@ def wrap_frozen_graph(gd, inputs, outputs):
         # Load external metadata YAML
         if isinstance(metadata, (str, Path)) and Path(metadata).exists():
             metadata = yaml_load(metadata)
+        else:
```
Comment (khatami-mehrdad) [inline ultralytics/nn/autobackend.py]: Explain this change.

File: ultralytics/nn/modules/conv.py [hunk 8]
```diff
@@ -53,7 +53,6 @@ def forward_fuse(self, x):
         """Perform transposed convolution of 2D data."""
         return self.act(self.conv(x))
```
Comment (khatami-mehrdad) [inline ultralytics/nn/modules/conv.py]: can you remove this change?

File: ultralytics/nn/modules/head.py [hunk 9]
```diff
@@ -179,13 +179,14 @@ def __init__(self, nc=80, kpt_shape=(17, 3), ch=()):
     def forward(self, x):
         """Perform forward pass through YOLO model and return predictions."""
         bs = x[0].shape[0]  # batch size
-        if self.export:
-            kpt = torch.cat([torch.permute(self.cv4[i](x[i]), (0, 2, 3, 1)).reshape(bs, -1, self.nk) for i in range(self.nl)], 1)
-        else:
-            kpt = torch.cat([self.cv4[i](x[i]).view(bs, self.nk, -1) for i in range(self.nl)], -1)  # (bs, 17*3, h*w)
+        kpt = torch.cat([self.cv4[i](x[i]).view(bs, self.nk, -1) for i in range(self.nl)], -1)  # (bs, 17*3, h*w)
```
Comment (khatami-mehrdad) [inline ultralytics/nn/modules/head.py]: why have you changed Detect forward?

File: ultralytics/utils/metrics.py [hunk 10]
```diff
@@ -1190,10 +1191,17 @@ def __init__(self) -> None:
         self.speed = {"preprocess": 0.0, "inference": 0.0, "loss": 0.0, "postprocess": 0.0}
         self.task = "classify"
 
-    def process(self, targets, pred):
+    def process(self, targets, pred, img_names):
```
Comment (khatami-mehrdad) [inline ultralytics/utils/metrics.py]: Is this changing all modes? Even detection?
img_names needed everywhere?

**Review by khatami-mehrdad — CHANGES_REQUESTED** [review]

File: ultralytics/cfg/__init__.py (hunk 1 above)
Comment (nklimov23) [reply ultralytics/cfg/__init__.py]: I can replace this with a 10-image IMDB dataset.

File: ultralytics/data/augment.py (hunk 2 above)
Comment (nklimov23) [reply ultralytics/data/augment.py]: Yes, they were helpful for extracting input tensors to the regression models, but they are not necessary now. I can remove them.

File: ultralytics/models/yolo/classify/val.py (hunk 5 above)
Comment (nklimov23) [reply ultralytics/models/yolo/classify/val.py]: No, the only changes in classify val are lines of code that save incorrectly classified image names to a file. This change does not affect regress val. Regress val has analogous code to save outlying image names to a file, and these changes in regress and classify were helpful for model accuracy analysis. I could remove them from being part of the pull request.

File: ultralytics/models/yolo/pose/val.py (hunk 6 above)
Comment (nklimov23) [reply ultralytics/models/yolo/pose/val.py]: This was part of an experiment, and I forgot to revert these changes, I apologize. I will change them back.

File: ultralytics/nn/modules/head.py (hunk 9 above)
Comment (nklimov23) [reply ultralytics/nn/modules/head.py]: This change is in Pose forward, and it was part of the same experiment with pose models mentioned above. I will change them back.

File: ultralytics/utils/metrics.py (hunk 10 above)
Comment (nklimov23) [reply ultralytics/utils/metrics.py]: No, this change is only specific to classification because this "process" method is called from the ClassificationValidator's "get_stats" method, and all of the arguments to this "process" method are specific to the ClassificationValidator.

File: ultralytics/nn/autobackend.py (hunk 7 above)
Comment (nklimov23) [reply ultralytics/nn/autobackend.py]: This was a temporary fix but not necessary anymore. I will remove it.

File: dg_export.py [hunk 11]
```diff
@@ -23,6 +22,6 @@ def parser_arguments():
 
     model = YOLO(args.weights)
 
-    success = model.export(format=args.format, imgsz=args.imgsz, data=args.data, int8=args.quantize, separate_outputs=True, export_hw_optimized=True, uint8_io_dtype=True, max_ncalib_imgs=args.max_ncalib_imgs)
+    success = model.export(format=args.format, simplify=True, imgsz=args.imgsz, data=args.data, int8=args.quantize, separate_outputs=True, export_hw_optimized=True, uint8_io_dtype=True, max_ncalib_imgs=100)
```
Comment (khatami-mehrdad) [inline dg_export.py]: can you add back this line?

    parser.add_argument('--max_ncalib_imgs', type=int, default=300, help='calibration image set size')	
  
make default = 100
